incubator/hnc/pkg/stats/metrics.go

```diff
@@ -27,18 +53,96 @@ var (
+	}
+
+	objectReconcileTotalView = &ocview.View{
+		Name:        "hnc/reconcilers/object/total",
+		Measure:     objectReconcileTotal,
+		Description: "The total number of object reconciliations happened",
+		Aggregation: ocview.LastValue(),
+		TagKeys:     []tag.Key{KeyGroupKind},
+	}
+
+	objectReconcileConcurrentView = &ocview.View{
+		Name:        "hnc/reconcilers/object/concurrent_peak",
+		Measure:     objectReconcileConcurrent,
+		Description: "The peak concurrent object reconciliations happened in the past 60s, which is also the minimum Stackdriver reporting period and the one we're using",
+		Aggregation: ocview.LastValue(),
+		TagKeys:     []tag.Key{KeyGroupKind},
+	}
+
+	objectWritesView = &ocview.View{
+		Name:        "hnc/reconcilers/object/object_writes_total",
+		Measure:     objectWritesTotal,
+		Description: "The number of object writes happened during object reconciliations",
+		Aggregation: ocview.LastValue(),
+		TagKeys:     []tag.Key{KeyGroupKind},
+	}
 )
 
+// periodicPeak contains periodic peaks for concurrent reconciliations.
+type periodicPeak struct {
+	// Lock is required for concurrent writes.
+	lock                          sync.Mutex
+	concurrentHierConfigReconcile counter
+	concurrentObjectReconcile     map[schema.GroupKind]counter
+}
+
+var peak periodicPeak
+
 func startRecordingMetrics() {
 	log := ctrl.Log.WithName("metricsRecorder")
 
-	// Register the view. It is imperative that this step exists,
+	// Register the views. It is imperative that this step exists,
 	// otherwise recorded metrics will be dropped and never exported.
-	if err := ocview.Register(hierReconcileView, hierWritesView); err != nil {
+	if err := ocview.Register(
+		hierReconcileTotalView,
+		hierReconcileConcurrentView,
+		hierWritesView,
+		namespaceWritesView,
+		objectReconcileTotalView,
+		objectReconcileConcurrentView,
+		objectWritesView,
+	); err != nil {
 		log.Error(err, "Failed to register the views")
 	}
+
+	// Start a new thread to record periodic peak.
+	go recordPeakConcurrentReconciles()
 }
 
-func recordMetricsInt64(c counter, ms *ocstats.Int64Measure) {
-	ocstats.Record(context.Background(), ms.M(int64(c)))
+// recordMetric records a measurement to a predefined measure without any specific tags.
+func recordMetric(m counter, ms *ocstats.Int64Measure) {
+	ocstats.Record(context.Background(), ms.M(int64(m)))
+}
+
+// recordTagMetric inserts a tag to the context before recording a metric.
+// The tag is used to group and filter collected metrics later on.
+func recordTagMetric(m counter, ms *ocstats.Int64Measure, k tag.Key, v string) {
+	ctx, _ := tag.New(context.Background(), tag.Insert(k, v))
+	ocstats.Record(ctx, ms.M(int64(m)))
+}
+
+func recordPeakConcurrentReconciles() {
+	for {
+		// This runs forever. It records and resets the peakConcurrent_ values every 1 minute,
+		// which is the same as the Stackdriver exporter's reporting interval.
+		time.Sleep(ReportingInterval)
+		peak.lock.Lock()
+		defer peak.lock.Unlock()
```

Comment:
`defer` is only invoked when the _function_ exits, not the _scope_. So I think you need an explicit call to Unlock here.